source/game_sa/CTagManager.cpp

```diff
@@ -0,0 +1,243 @@
+    ReversibleHooks::Install("CTagManager", "SetAlpha_Entity", 0x49CEC0, (void(*)(CEntity*, unsigned char))(&CTagManager::SetAlpha));
+    ReversibleHooks::Install("CTagManager", "GetNearestTag", 0x49D160, &CTagManager::GetNearestTag);
+    ReversibleHooks::Install("CTagManager", "SetupAtomic", 0x49CE10, &CTagManager::SetupAtomic);
+    ReversibleHooks::Install("CTagManager", "RenderTagForPC", 0x49CE40, &CTagManager::RenderTagForPC);
+    ReversibleHooks::Install("CTagManager", "Save", 0x5D3D60, &CTagManager::Save);
+    ReversibleHooks::Install("CTagManager", "Load", 0x5D3DA0, &CTagManager::Load);
+}
+
+void CTagManager::Init()
+{
+    CTagManager::ms_numTags = 0;
+    CTagManager::ms_numTagged = 0;
+}
+
+void CTagManager::ShutdownForRestart()
+{
+    for (int32_t i = CTagManager::ms_numTags; i > 0; --i)
+        CTagManager::ms_tagDesc[i].m_nAlpha = 0;
+
+    CTagManager::ms_numTagged = 0;
+}
+
+CVector& CTagManager::GetTagPos(int iTag)
+{
+    return CTagManager::ms_tagDesc[iTag].m_pEntity->GetPosition();
+}
+
+void CTagManager::AddTag(CEntity* pEntity)
+{
+    auto& pTag = CTagManager::ms_tagDesc[CTagManager::ms_numTags];
+    pTag.m_pEntity = pEntity;
+    pTag.m_nAlpha = 0;
+
+    ++CTagManager::ms_numTags;
+}
+
+tTagDesc* CTagManager::FindTagDesc(CEntity* pEntity)
+{
+    if (!CTagManager::ms_numTags)
+        return nullptr;
+
+    for (int32_t i = CTagManager::ms_numTags - 1; i >= 0; --i) {
+        auto pTagDesc = &CTagManager::ms_tagDesc[i];
+        if (pTagDesc->m_pEntity == pEntity)
+            return pTagDesc;
+    }
+
+    return nullptr;
+}
+
+bool CTagManager::IsTag(CEntity const* pEntity)
+{
+    auto pModelInfo = CModelInfo::GetModelInfo(pEntity->m_nModelIndex);
+    if (pModelInfo->GetRwModelType() != rpATOMIC)
+        return false;
+
+    return pModelInfo->IsTagModel() && !pModelInfo->AsAtomicModelInfoPtr()->bTagDisabled;
+}
+
+void CTagManager::SetAlpha(RpAtomic* pAtomic, unsigned char ucAlpha)
+{
+    CVisibilityPlugins::SetUserValue(pAtomic, ucAlpha);
+}
+
+unsigned char CTagManager::GetAlpha(RpAtomic* pAtomic)
+{
+    return CVisibilityPlugins::GetUserValue(pAtomic);
+}
+
+unsigned char CTagManager::GetAlpha(CEntity* pEntity)
+{
+    if (pEntity->m_pRwAtomic)
+        return CVisibilityPlugins::GetUserValue(pEntity->m_pRwAtomic);
+
+    auto pTag = CTagManager::FindTagDesc(pEntity);
+    assert(pTag); // Originally the function would access unitialized memory, by clearing EAX and dereferencing pointer to [EAX + 0x4] right after that
+
+    return pTag->m_nAlpha;
+}
+
+void CTagManager::ResetAlpha(CEntity* pEntity)
+{
+    if (!pEntity->m_pRwAtomic)
+        return;
+
+    auto pTagDesc = CTagManager::FindTagDesc(pEntity);
+    CTagManager::SetAlpha(pEntity->m_pRwAtomic, pTagDesc->m_nAlpha);
+}
+
+int64_t CTagManager::GetPercentageTagged()
+{
+    return static_cast<int64_t>(static_cast<double>(CTagManager::ms_numTagged) / static_cast<double>(CTagManager::ms_numTags) * 100.0F);
```

Comment:
```suggestion
{
    return static_cast<int64_t>(static_cast<float>(CTagManager::ms_numTagged) / static_cast<float>(CTagManager::ms_numTags) * 100.0F);
```
Always use floating points. R* devs accidentally used doubles here. I won't go into too much details, but let's say floating points are also good performance-wise.

Reply:
In this case it actually wont make any difference i think, the values never leaver fpu stack, it's stored as 64 bit integer straight after the calculations. This function is a bit weird, in such that it ends with jmp to ftol procedure, it doesn't modify any registers or anything, you could double check if i didn't make some mistake trying to understand it :D

Reply:
Okay, this part of code is fine because I don't see any warnings in VS, but you forgot casts in other parts of code. In fact, you don't even need to cast from double to float or vice-versa if you just use floats everywhere, this would get rid of the warnings.

Reply:
In case of those 2 percentage functions casting to float introduces warnings, cuz the return is 64 bit, imma leave these at that, and fix places with missing float suffix
![obraz](https://user-images.githubusercontent.com/4185198/103791327-9c94e780-5042-11eb-9c18-3b7160c6587f.png)

Reply:
Fixed